### DISCUSSION TITLE
loadPyodide loading progress

### DISCUSSION THREAD

**Discussion by fcrozatier** [pyodide/pyodide#2927] [Q&A] [answered]
How can I know the loading progress of loadPyodide? At the moment loadPyodide returns a promise so it's not obvious to me how to do this.

Since it takes quite some time to load, wouldn't it be a nice feature to have ? This would let users know something is happening in the background, especially useful on slower networks.

**Comment by ryanking13** [comment, marked as answer]
There is no "official" way to get the progress for now. It is because most time taking part in loading Pyodide is fetching and compiling `pyodide.asm.wasm`, but that part is mostly auto generated from the Emscripten compiler, so it is hard to inject a custom code there.

One hacky way of getting the progress is [hacking global `fetch` function](https://stackoverflow.com/questions/65491241/get-webassembly-instantiatestreaming-progress) such as:

```js
let fetchOriginal = fetch;
let fetchWithProgress = async (url) => {
	if (typeof url !== "string" || !url.includes("pyodide.asm.wasm")) {
		return fetchOriginal(url);
	}

	const response = await fetchOriginal(url);
	// May be incorrect if compressed
	const contentLength = response.headers.get("Content-Length");
	const total = parseInt(contentLength, 10);

	let bytesLoaded = 0;
	const ts = new TransformStream({
	  transform (chunk, ctrl) {
                // Put some custom progress function here.
		bytesLoaded += chunk.byteLength;
		console.log(bytesLoaded / total);
		ctrl.enqueue(chunk)
	  }
	});

	return new Response(response.body.pipeThrough(ts), response);
}
globalThis.fetch = fetchWithProgress;

await loadPyodide(...);

globalThis.fetch = fetchOriginal; // put back original fetch
```

**Reply by hoodmane** [reply]
@sbc100

**Reply by fcrozatier** [reply]
It's a great hack thanks !

One improvement would be the precision of `contentLength` because  at the moment with pyodide 0.20.0 the uncompressed content length is roughly 2.9 times bigger than the compressed one. But this may change with future versions. Any ideas how to go about this?

Otherwise could we implement a custom header to get the uncompressed `contentLength`?

I believe any user feedback during the load / compile of pyodide is of great value considering the non negligeable time it can take.

**Reply by ryanking13** [reply]
> One improvement would be the precision of `contentLength` because at the moment with pyodide 0.20.0 the uncompressed content length is roughly 2.9 times bigger than the compressed one. But this may change with future versions. Any ideas how to go about this?

To be honest, I copied the code from the link so I'm not sure of it. It would be great if you can share us your findings.

> Otherwise could we implement a custom header to get the uncompressed contentLength?

Probably yes. However, adding a header is server-side work, so I don't think it is our job to do that. Maybe you can ask CDN providers or you'll need to make your own delivering server.

**Reply by fcrozatier** [reply]
> ... It would be great if you can share us your findings.

When implementing the above script, it turns out that the processed chunk add up to more than `contentLength` because the `Content-Length` header gets the compressed file size. So in the end the ratio `bytesLoaded / total` is more than 1, it's around 2.902.

To have a progress bar from 0 to about 1 I need to use magic numbers like `const contentLength = +response.headers.get("Content-Length") * 2.9` to approximate the uncompressed file size.

So it would be great to get this info from a custom header. 

> ... adding a header is server-side work, so I don't think it is our job to do that

If we think it's useful we could ask a maintainer who has access to the CDN if it is possible to add this custom header.

> Maybe you can ask CDN providers or you'll need to make your own delivering server.

Seems unreasonnable to go down that way for a custom header.

**Reply by hoodmane** [reply]
Bizarre that there isn't a standard way to do this...

**Reply by fcrozatier** [reply]
Issue [#2947](https://github.com/pyodide/pyodide/issues/2947)

**Reply by hoodmane** [reply]
Thanks @fcrozatier
